packages/zoltan2/test/connectivity/iceSheetIntegration.cpp

```diff
@@ -0,0 +1,227 @@
+  
+
+  Teuchos::RCP<SparseMatrix> Matrix;
+  Matrix = uinput->getUITpetraCrsMatrix();
+  
+  if(prepartition){  
+    //partition the input matrix, to test noncontiguous vertex ID distributions
+    SparseMatrixAdapter *zadapter;
+    zadapter = new SparseMatrixAdapter(Matrix,1);
+    zadapter->setRowWeightIsNumberOfNonZeros(0);
+
+    Teuchos::ParameterList zparams;
+    zparams.set("algorithm","parmetis");
+    zparams.set("imbalance_tolerance",1.05);
+    zparams.set("partitioning_approach","partition");
+    Zoltan2::PartitioningProblem<SparseMatrixAdapter> zproblem(zadapter, &zparams);
+    zproblem.solve();
+   
+    // print partition characteristics before and after
+    typedef Zoltan2::EvaluatePartition<SparseMatrixAdapter> quality_t;
+    quality_t evalbef(zadapter, &zparams, comm, NULL);
+    if(me == 0){
+      std::cout << "BEFORE PREPARTITION: Partition statistics:" <<std::endl;
+      evalbef.printMetrics(std::cout);
+    }
+
+    quality_t evalaft(zadapter, &zparams, comm, &zproblem.getSolution());
+    if(me == 0) {
+      std::cout << "AFTER PREPARTITION: Partition statistics:" << std::endl;
+      evalaft.printMetrics(std::cout);
+    }
+
+    Teuchos::RCP<SparseMatrix> newMatrix;
+    zadapter->applyPartitioningSolution(*Matrix, newMatrix, zproblem.getSolution());
+  
+    Matrix = newMatrix;
+    delete zadapter;
+  }
+  
+  Teuchos::RCP<const CrsGraph> crsgraph = Matrix->getCrsGraph();  
+
+  Teuchos::RCP<GraphAdapter> inputGraphAdapter = rcp(new GraphAdapter(crsgraph));
+  
+
+  //need to read in problem specific files here, start out with zeroed out arrays
+  int nlocal = (int) inputGraphAdapter->getLocalNumVertices();
+  bool* basalFriction = new bool[nlocal];
+  const gno_t* vtxIDs = new gno_t[nlocal];
+  inputGraphAdapter->getVertexIDsView(vtxIDs);
+
+  std::cout<<me<<": num local vtxIDs = "<<nlocal<<"\n";
+
+  int* grounded_flags_global;
+  gno_t* boundary_edges_global;
+
+  int nglobal = 0;
+  int num_global_boundary_edges = 0;
+  if(me == 0){
+    read_grounded_file(argv[3],nglobal,grounded_flags_global);
+    read_boundary_file<gno_t>(argv[4],num_global_boundary_edges,boundary_edges_global);
+    
+  }
+  
+  //broadcast global array counts
+  Teuchos::broadcast<int,int>(*comm, 0,1,&nglobal);
+  Teuchos::broadcast<int,int>(*comm, 0,1,&num_global_boundary_edges);
+  
+  if(me != 0){
+    grounded_flags_global = new int[nglobal];
+    boundary_edges_global = new gno_t[num_global_boundary_edges];
+  }
+
+  //broadcast the global arrays, to trim them down to local
+  Teuchos::broadcast<int,int>(*comm,0,nglobal,grounded_flags_global);
+  Teuchos::broadcast<int,gno_t>(*comm,0,num_global_boundary_edges, boundary_edges_global);
+ 
+  int numLocalBoundaryEdges = 0;
+  for(int i = 0; i < num_global_boundary_edges; i+=2){
+    for(int j = 0; j < nlocal; j++){
+      if(boundary_edges_global[i] == vtxIDs[j] || boundary_edges_global[i+1] == vtxIDs[j]){
```

Comment:
Rather than linear search, how about using Map's check for global IDs?  Map->getLocalElement(boundary_edges_global[i]) != Teuchos::OrdinalTraits<LocalOrdinal>::invalid() ==> processor has the vtx
https://docs.trilinos.org/dev/packages/tpetra/doc/html/classTpetra_1_1Map.html#a11c6a4585d616718eeb5c16d8a492511

Reply:
I'll use crsgraph->getRowMap() as the map in both cases.

Reply:
I'll change this to use crsgraph->getRowMap() as the map for the ownership check.